### PR TITLE
*: add support for socket options

### PR DESCRIPTION
**Pull request by hexfusion**
This PR adds support for setting socket options `SO_REUSEADDR` and `SO_REUSEPORT` to etcd listeners via `ListenConfig`. These options give the flexibility to cluster admins who wish to more explicit control of these features. What we have found is during etcd process restart there can be a considerable time waiting for the port to release as it is held open by `TIME_WAIT` which on many systems is 60s.

```
--socket-reuse-port: enables SO_REUSEPORT [1]  allows rebind of a port already in use. 
--socket-reuse-address: enables SO_REUSEADDR [1]  allows binding to an address in `TIME_WAIT` state. 
```
```
$ ps aux | grep etcd | wc -l
0

$  ss --numeric -o state time-wait 
tcp    0       0                 10.0.138.12:34642            10.0.138.12:2380   timer:(timewait,50sec,0)                                                       
tcp    0       0                 10.0.138.12:34470            10.0.138.12:2380   timer:(timewait,35sec,0)                                                       
tcp    0       0                 10.0.138.12:34086            10.0.138.12:2380   timer:(timewait,10sec,0)                                                       
tcp    0       0                 10.0.138.12:34232            10.0.138.12:2380   timer:(timewait,20sec,0)                                                       
tcp    0       0                 10.0.138.12:34694            10.0.138.12:2380   timer:(timewait,55sec,0)                                                       
tcp    0       0                 10.0.138.12:34332            10.0.138.12:2380   timer:(timewait,25sec,0)                                                       
tcp    0       0                 10.0.138.12:33966            10.0.138.12:2380   timer:(timewait,845ms,0)                                                       
tcp    0       0                 10.0.138.12:34040            10.0.138.12:2380   timer:(timewait,5.845ms,0)                                                     
tcp    0       0                 10.0.138.12:34184            10.0.138.12:2380   timer:(timewait,15sec,0)                                                       
tcp    0       0                 10.0.138.12:34570            10.0.138.12:2380   timer:(timewait,45sec,0)                                                       
tcp    0       0                 10.0.138.12:34512            10.0.138.12:2380   timer:(timewait,40sec,0)                                                       
tcp    0       0                 10.0.138.12:34398            10.0.138.12:2380   timer:(timewait,30sec,0)  
```
So we can wait for many seconds even after etcd process is long dead for client and peer ports to become available.

```
2021-02-19 13:57:25.117257 C | etcdmain: listen tcp 10.0.138.12:2380: bind: address already in use
```

A similar approach has been taken with the addition of these options in the kube-apiserver[2]

[1] https://man7.org/linux/man-pages/man7/socket.7.html
[2] https://github.com/kubernetes/kubernetes/pull/93861

### CODE REVIEW

File: pkg/transport/timeout_listener.go [hunk 1]
```diff
@@ -23,19 +23,32 @@ import (
...
 	}
-	ln = &rwTimeoutListener{
+	return newTimeoutListener(ln, scheme, rdtimeoutd, wtimeoutd, tlsinfo)
+}
+
+// NewTimeoutListerWithSocketOpts returns a listener that listens on the given address.
+// If read/write on the accepted connection blocks longer than its time limit,
+// it will return timeout error. Socket options can be passed and will be applied to the
+// ListenConfig.
+func NewTimeoutListerWithSocketOpts(addr string, scheme string, tlsinfo *TLSInfo, rdtimeoutd, wtimeoutd time.Duration, sopts *SocketOpts) (net.Listener, error) {
```
Comment (gyuho) [inline pkg/transport/timeout_listener.go:37]: This function parameters are getting too verbose. Can we make it variadic with something like `ListenerOption`? Then we can get rid of all those `tlsinfo` and others.

https://github.com/etcd-io/etcd/blob/3d7aac948b1cde9de96f24363c8a790883d56c1d/client/v3/op.go#L333-L339

File: server/etcdmain/help.go [hunk 2]
```diff
@@ -85,6 +85,10 @@ Member:
     Frequency duration of server-to-client ping to check if a connection is alive (0 to disable).
   --grpc-keepalive-timeout '20s'
     Additional duration of wait before closing a non-responsive connection (0 to disable).
+  --socket-reuse-port 'false'
+    Enable to set socket option SO_REUSEPORT on listeners allowing rebinding of a port already in use.
+  --socket-reuse-address 'false'
+	Enable to set socket option SO_REUSEADDR on listeners allowing binding to an address in TIME_WAIT state.
```
Comment (gyuho) [inline server/etcdmain/help.go:91]: Fix indentation? :)

**Comment by gyuho** [comment]
nice find!

File: pkg/transport/timeout_listener.go (hunk 1 above)
Comment (hexfusion) [reply pkg/transport/timeout_listener.go:37]: make sense thanks

**Comment by gyuho** [comment]
@hexfusion Any updates? Would like to get this in release 3.5

**Comment by pires** [comment]
cc @jdef

**Comment by hexfusion** [comment]
> @hexfusion Any updates? Would like to get this in release 3.5

@gyuho I will have an update early next week latest

**Comment by hexfusion** [comment]
@gyuho PTAL added fixup commit will rebase once the review is done.

**Review by gyuho — APPROVED** [review]

**Comment by gyuho** [comment]
@hexfusion Thanks! Can we also update our 3.5 CHANGELOG?

**Comment by Quentin-M** [comment]
This is amazing, thanks for doing this!

**Comment by vivekpatani** [comment]
Hi, the integration test cluster seems to not have access to this option, was that deliberate or is it something that we should consider adding? @serathius @ahrtr @gyuho

**Comment by ahrtr** [comment]
> Hi, the integration test cluster seems to not have access to this option, was that deliberate or is it something that we should consider adding? @serathius @ahrtr @gyuho

Please try to add common config items so that they can be reused by both integration and e2e test cases.

**Comment by vivekpatani** [comment]
Sounds good, I'll work on this, since when I tried working on tests for my other patch (https://github.com/etcd-io/etcd/issues/14355), and I was restarting the cluster, I ran into this issue, would be useful to have. Thanks, will take this up.

**Pull request merged**